addon/@ember/test-waiters/wait-for.ts

```diff
@@ -96,3 +135,33 @@ function isThenable(
     typeof maybePromise.then === 'function'
   );
 }
+
+function isGenerator(
+  maybeGenerator: any | CoroutineGenerator<unknown>
+): maybeGenerator is CoroutineGenerator<unknown> {
+  return typeof maybeGenerator[Symbol.iterator] === 'function';
+}
+
+const GENERATOR_WAITER = buildWaiter('@ember/test-waiters:generator-waiter');
+
+function waitForGenerator<T>(
+  generator: CoroutineGenerator<T>,
+  label?: string
+): CoroutineGenerator<T> {
+  GENERATOR_WAITER.beginAsync(generator, label);
+
+  return {
+    next() {
+      try {
+        let val = generator.next();
+        if (val.done) {
+          GENERATOR_WAITER.endAsync(generator);
+        }
+        return val;
+      } catch (e) {
+        GENERATOR_WAITER.endAsync(generator);
+        throw e;
+      }
+    },
+  } as CoroutineGenerator<T>;
+}
```

Comment:
This seems vaguely wrong to my eyes (though maybe the coffee just hasn't kicked in yet). We basically need to follow these semantics:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#The_iterator_protocol

As far as I can tell we have a few problems with the current implementation:

* we cannot return `val` from `next` (we have to **always** return our custom iterable, or we don't catch subsequent async in follow up `next` calls)
* We should bracketing the start/stop of async around calling the original iterables `next` (we can't start once before the whole thing, then end after the first next)
* our return value does not have `done` (but AFAICT it must)

Reply:
Either you're missing something or I am 😆 I think the naming is confusing -- a [Generator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator) (object) *is* an iterator that is returned by a [Generator Function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*). And it's also an iterable. `ember-concurrency` only uses the iterator interface not the iterable interface, which I think might be the source of your confusion?

When we called the function that we're wrapping [here](https://github.com/emberjs/ember-test-waiters/pull/173/files#diff-41ec033b81ca85a4a578fb8082225f03R116), if it was a generator function, we got back a generator object, so that's what we're dealing with here. We're proxying the iterator (not iterable) protocol, which only consists of a `next()` method, so we're returning an object that also implements the iterator protocol, and whose `next()` method just forwards to the underlying/wrapped generator object's `next()` method, but watches for its return value indicating that the iteration/generation is done, or for it throwing an error, so it can stop the waiter. So the ordering here is:

1. We call the generator function to get back a generator object/iterator
2. We start test-waiting
3. We return a wrapper object implementing the iterator procotol
4. Sometime later, someone calls `next()` on our wrapper object, we call `next()` on the generator object, it returns a result saying it's not done iterating, and we return that result
5. One of:
   * Sometime later, someone calls `next()` again on our wrapper object, we call `next()` on the generator object, it returns a result saying it is done iterating, we stop test-waiting and return the result
   * Sometime later, someone calls `next()` again on our wrapper object, we call `next()` on the generator, it throws an error, we stop test-waiting and re-throw the error

I'm pretty sure this is all nominally correct, although I've realized a few things from being challenged (the internet is hard, so just want to emphasize that in this context I see being challenged as a 100% good thing!) and thinking through it again:

1. Generator objects also implement `return()` and `throw()`. I think `ember-concurrency` uses `throw()` for task cancellation? Anyway, we should definitely proxy those methods onto our wrapper object as well.
2. While I don't think `ember-concurrency` does this, callers can continue calling `next()` on generator objects after the iteration has completed, and they'll continue getting `{ value: undefined, done: true }`. In this case we would continue calling `endAsync()` on the waiter, which is very bad, so I suppose we probably need an `isDone` function-scoped variable to track when the generator has run to completion and we've already ended test-waiting so we don't do it again.
3. A generator object's `next()` accepts an argument that shows up in the generator function as a return value from the previous `yield` statement -- I'm pretty certain `ember-concurrency` never does this, but I suppose we should forward arguments from our wrapper's `next()` function to the generator object's `next()` function anyway
4. Since generator objects also implement the iterable protocol, we could implement `[Symbol.iterator]` on our proxy object and have it just return itself (the proxy object).
5. We should definitely write some tests covering basic `ember-concurrency` functionality (multi-`yield` task functions, throwing errors from the task function, task cancellation, etc.) to make sure our wrapping doesn't interfere with anything.
6. We should *maybe* write some generator object interface tests to unit test that our wrapper object doesn't interfere with the generator object interface? I'm not really sure how far we need to go with this right now since currently our only known use case is `ember-concurrency`.

Some of these definitely make sense to do now. But it's not clear to me how much effort it makes sense to spend right now on cases that don't happen in `ember-concurrency`.

Let me know your thoughts.